### PR TITLE
Create logger instance

### PR DESCRIPTION
**Pull request by dkryvi**
#3

### CODE REVIEW

File: app/services/logger.js [hunk 1]
```diff
@@ -0,0 +1,43 @@
+const bunyan = require('bunyan');
+const PrettyStream = require('bunyan-prettystream');
+const { get } = require('lodash');
+
+const Logger = require('./base_logger');
+
+const pkg = require('../../package.json');
```
Comment (Vanish404) [inline app/services/logger.js]: Why we need to import `package.json`? Is it correct to chose between `config.name` and `pkg.name`?

File: app.js [hunk 2]
```diff
@@ -1,6 +1,9 @@
 const initializer = require('./app/utils/initializer');
+const initilizeGracefulExit = require('./app/utils/initialize_graceful_exit');
```
Comment (akosenko) [inline app.js]: please provide some info why we should use it?

File: app/server.js [hunk 3]
```diff
@@ -2,6 +2,8 @@ const app = require('express')();
 const nconf = require('nconf');
 const globSync = require('glob').sync;
 
+const logger = require('./services/logger');
```
Comment (akosenko) [inline app/server.js]: are you sure that logger is a service?

File: app/services/logger.js (hunk 1 above)
Comment (akosenko) [reply app/services/logger.js]: wow-wow, package.json!?

File: app/services/logger.js [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+  const prettyStream = new PrettyStream();
+  prettyStream.pipe(process.stdout);
+
+  loggerOptions.streams.push({
+    level: 'info',
+    stream: prettyStream,
+  });
+}
+
+const rootLogger = new Logger(bunyan.createLogger(loggerOptions));
```
Comment (akosenko) [inline app/services/logger.js]: `bunyan.createLogger(loggerOptions)` creates parent logger, so I'm not sure that we should wrap them into our own additional base_logger

File: app/services/logger.js [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+
+const loggerOptions = {
+  name: config.name || pkg.name,
+  streams: [
+    { level: 'error', stream: process.stderr },
+    { level: 'error', path: 'logs/error.log' },
+    ...get(config, 'streams', [])
+  ],
+  serializers: bunyan.stdSerializers,
+};
```
Comment (akosenko) [inline app/services/logger.js]: Strange construction for logger. Please create correct logger configuration in our config file, for example:
{
"log": {
    "name": "event-emitter",
    "streams": [{
      "level": "debug",
      "path": "logs/event-emitter.log",
       period: '1d',
       count: 3
    }]
  }
}

File: app/services/logger.js [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+
+if (config.rotationEnabled) {
+  loggerOptions.streams.push({
+    type: 'rotating-file',
+    path: `logs/${env}.log`,
+    period: '1d',
+    count: 7
+  });
+}
```
Comment (akosenko) [inline app/services/logger.js]: move it to config file

File: app/utils/initialize_graceful_exit.js [hunk 7]
```diff
@@ -0,0 +1,12 @@
+const config = require('../config');
+const logger = require('../services/logger');
+
+module.exports = () => {
+  process.once('SIGUSR2', () => {
+    process.kill(process.pid, 'SIGUSR2');
```
Comment (akosenko) [inline app/utils/initialize_graceful_exit.js]: why you kill process here?

**Review by akosenko — CHANGES_REQUESTED** [review]

File: app/services/logger.js (hunk 1 above)
Comment (akosenko) [reply app/services/logger.js]: wow-wow, package.json!?

File: app/services/logger.js (hunk 5 above)
Comment (akosenko) [inline app/services/logger.js]: Strange construction for logger. Please create correct logger configuration in our config file, for example:
{
"log": {
"name": "event-emitter",
"streams": [{
"level": "debug",
"path": "logs/event-emitter.log",
period: '1d',
count: 3
}]
}
}

File: app/services/logger.js (hunk 6 above)
Comment (akosenko) [inline app/services/logger.js]: move it to config file

**Review by akosenko — CHANGES_REQUESTED** [review]

File: app/server.js (hunk 3 above)
Comment (akosenko) [inline app/server.js]: are you sure that logger is a service?

**Review by akosenko — CHANGES_REQUESTED** [review]

File: app/utils/initialize_graceful_exit.js (hunk 7 above)
Comment (dkryvi) [reply app/utils/initialize_graceful_exit.js]: I think in this place later we can use this [library](https://github.com/emostar/express-graceful-exit) or write something else to grace stop our server or close all socket connections in the app.  Maybe it doesn't the best place listen to this event. I'm thinking of moving the code to restart logger streams to the logger file, but where I should listen `SIGUSR2` to kill the process?

File: app/services/logger.js (hunk 6 above)
Comment (dkryvi) [reply app/services/logger.js]: [Done](https://github.com/inst-event-emitter/eventemitter-backend/pull/7/commits/44502c79cf19007b92f9f6739774ab39f2674789#diff-a18e55398c3a9a4054717c2ccfa889fdR11)

File: app/services/logger.js (hunk 5 above)
Comment (dkryvi) [reply app/services/logger.js]: [Done](https://github.com/inst-event-emitter/eventemitter-backend/pull/7/commits/44502c79cf19007b92f9f6739774ab39f2674789#diff-2f876aef44739dc3746ddb1485ae379fR15)

File: app/services/logger.js (hunk 1 above)
Comment (dkryvi) [reply app/services/logger.js]: [Fixed](https://github.com/inst-event-emitter/eventemitter-backend/pull/7/commits/44502c79cf19007b92f9f6739774ab39f2674789#diff-2f876aef44739dc3746ddb1485ae379fR12)

File: app/services/base_logger.js [hunk 8]
```diff
@@ -0,0 +1,43 @@
+class Logger {
+  constructor(bunyanLogger) {
+    this.bunyanLogger = bunyanLogger;
+    this.children = [];
+  }
+
+  child(opts) {
```
Comment (Vanish404) [inline app/services/base_logger.js]: In witch cases do we need to call a child logger? Would you might providing some information about that?

File: app/services/logger.js [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+  const prettyStream = new PrettyStream();
+  prettyStream.pipe(process.stdout);
+
+  loggerOptions.streams.push({
+    level: 'info',
+    stream: prettyStream,
+  });
+}
+
+const rootLogger = new Logger(bunyan.createLogger(loggerOptions));
```
Comment (Vanish404) [inline app/services/logger.js]: suppose that we deleted the folder for logging or we haven't got her, do we need to check this case for a error? or logger instance create a new folder for logs, don't it?

Comment (DenKrivichanin) [reply app/services/logger.js]: image
In this case, child logger is created with the specific name and when writing f.e `log.info` in the console we can see this `component` field and understand were this log was emitted

File: app/services/base_logger.js (hunk 8 above)
Comment (dkryvi) [reply app/services/base_logger.js]: image
In this case, child logger is created with the specific name and when writing f.e `log.info` in the console we can see this `component` field and understand were this log was emitted

File: app/services/logger.js (hunk 9 above)
Comment (dkryvi) [reply app/services/logger.js]: I checked this case and error has occurred
image

Comment (dkryvi) [reply app/services/logger.js]: In this case we can check if folder exists before create logger
```
const fs = require('fs');
const dir = './logs';

if (!fs.existsSync(dir)){
    fs.mkdirSync(dir);
}
```

File: app/services/base_logger.js (hunk 8 above)
Comment (DenKrivichanin) [reply app/services/base_logger.js]: You can check it. [Link](https://github.com/inst-event-emitter/eventemitter-backend/pull/7/commits/ad112cb8c1d9c64199e855f153ea3258174bfc77#diff-586cf549a824167942b3b214a6303c48R29)

File: app/services/logger.js (hunk 9 above)
Comment (dkryvi) [reply app/services/logger.js]: You can check it. [Link](https://github.com/inst-event-emitter/eventemitter-backend/pull/7/commits/ad112cb8c1d9c64199e855f153ea3258174bfc77#diff-586cf549a824167942b3b214a6303c48R29)

File: app/server.js (hunk 3 above)
Comment (dkryvi) [reply app/server.js]: I moved logger to `utils` folder. See [commit](https://github.com/inst-event-emitter/eventemitter-backend/pull/7/commits/e02b618176eba1d6fdcc919149f268e3cea63629#diff-f53e59f241bb62f0261e245350683e07R5)

**Review by akosenko — APPROVED** [review]

**Review by akosenko — APPROVED** [review]

File: app/config/dev.json [hunk 10]
```diff
@@ -1,13 +1,7 @@
 {
   "logger": {
-    "name": "event-emitter",
-    "logsDir": "logs",
     "streams": [{
-      "level": "error",
-      "path": "error.log"
-    }, {
-      "type": "rotating-file",
```
Comment (dkryvi) [inline app/config/dev.json]: Why you remove `rotating-file` stream?

File: app/utils/logger.js [hunk 11]
```diff
@@ -1,52 +1,43 @@
...
-
-    return stream;
-  }),
-  serializers: bunyan.stdSerializers,
-}));
-
-process.once('SIGUSR2', () => {
-  rootLogger.reopenFileStreams();
+const parent = bunyan.createLogger(logConfig);
+const loggerCollector = [parent];
```
Comment (dkryvi) [inline app/utils/logger.js]: Why should we store all loggers in the array?

File: app/utils/logger.js [hunk 12]
```diff
@@ -1,52 +1,43 @@
...
+// Process "reloadLogs" pm2 command, use new .log file after log rotation
+process.on('SIGUSR2', () => {
+  loggerCollector.forEach(logger => logger.reopenFileStreams());
```
Comment (dkryvi) [inline app/utils/logger.js]: Should we reopen streams for all loggers or only for child loggers(like in previous realization)?

File: app/utils/logger.js [hunk 13]
```diff
@@ -1,52 +1,43 @@
...
-module.exports = rootLogger;
+const getLogger = (component, options) => {
+  const logger = parent.child(Object.assign({
+    component: component || 'Undefined component'
+  }, options));
+  loggerCollector.push(logger);
+  return logger;
+};
+
+module.exports = getLogger;
```
Comment (dkryvi) [inline app/utils/logger.js]: Please, add new line

File: app/utils/logger.js (hunk 11 above)
Comment (akosenko) [reply app/utils/logger.js]: to reopen them with help of reopenFileStreams

File: app/utils/logger.js (hunk 12 above)
Comment (akosenko) [reply app/utils/logger.js]: please have a look on previous realization
reopenFileStreams() {
    this.bunyanLogger.reopenFileStreams();
    this.children.forEach(c => c.reopenFileStreams()); 
}
this method reopen streams for current root log and for all his childs(that also stored in the array)

File: app/utils/logger.js (hunk 13 above)
Comment (akosenko) [reply app/utils/logger.js]: done

File: app/utils/logger.js (hunk 12 above)
Comment (dkryvi) [reply app/utils/logger.js]: Ok. Should we have comment line about pm2?

Comment (akosenko) [reply app/utils/logger.js]: done

**Review by akosenko — APPROVED** [review]

File: app/utils/logger.js (hunk 12 above)
Comment (dkryvi) [reply app/utils/logger.js]: GLTM. I think we can merge it pr

**Pull request merged**